### PR TITLE
Adding a minimal implementation of NeRF

### PR DESCRIPTION
**Pull request by ariG23498**
In this PR we have implemented a minimal version of [NeRF](https://arxiv.org/abs/2003.08934). This paper deals with novel view generation of a 3-D scene using a sparse set of 2-D images. We have tried wrapping the model into a `keras.Model` to utilise callbacks to visualize the entire training step. We think this example would be a starting point for model implementation in the field of computer graphics and 3-D rendering.

[Colab](https://colab.research.google.com/drive/1N_0fnGBfLS9XkYxboTtZlBceqs9Bw7pk?usp=sharing)

### CODE REVIEW

File: examples/vision/nerf.py [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+color."
+
+Here we have gone for a minimal implementation and have used `64`
+Dense units instead of `256` as that mentioned in the paper.
+"""
+
+
+def get_nerf_model(num_layers, num_pos):
+    # Initialize the input layer.
+    inputs = tf.keras.Input(shape=(num_pos, 2 * 3 * POS_ENCODE_DIMS + 3))
```
Comment (fchollet) [inline examples/vision/nerf.py:348]: Your lines will be marginally shorter if you import `keras` and the `layers` module, then access members of these modules instead of `tf`

File: examples/vision/nerf.py [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+    else:
+        delta = tf.concat(
+            [delta, tf.broadcast_to([1e10], shape=(BATCH_SIZE, 1))], axis=-1
+        )
+        alpha = 1.0 - tf.exp(-sigma_a * delta[:, None, None, :])
+
+    # Get transmittance.
+    exp_term = 1.0 - alpha
+    epsilon = 1e-10
+    T = tf.math.cumprod(exp_term + epsilon, axis=-1, exclusive=True)
```
Comment (fchollet) [inline examples/vision/nerf.py:396]: Use lowercase names.

File: examples/vision/nerf.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+    outputs = tf.keras.layers.Dense(units=4)(x)
+
+    # Create the model.
+    model = tf.keras.Model(inputs=inputs, outputs=outputs)
+
+    # Return the model.
+    return model
+
+
+def render_rgb_depth(model, rays_flat, t_vals, rand=True, train=True):
```
Comment (fchollet) [inline examples/vision/nerf.py:367]: Add a short docstring.

File: examples/vision/nerf.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+
+    if rand:
+        depth_map = tf.reduce_sum(weights * t_vals, axis=-1)
+    else:
+        depth_map = tf.reduce_sum(weights * t_vals[:, None, None], axis=-1)
+    return (rgb, depth_map)
+
+
+"""
+# "Model.fit"
```
Comment (fchollet) [inline examples/vision/nerf.py:408]: Training

File: examples/vision/nerf.py [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+        gradients = tape.gradient(loss, trainable_variables)
+
+        # Apply the grads and optimize the model.
+        self.optimizer.apply_gradients(zip(gradients, trainable_variables))
+
+        # Get the PSNR of the reconstructed images and the source images.
+        psnr = tf.image.psnr(images, rgb, max_val=1.0)
+
+        # return the loss and the psnr value
+        return {"loss": loss, "psnr": psnr}
```
Comment (fchollet) [inline examples/vision/nerf.py:450]: To track the loss and metric, use `Mean` metric trackers (otherwise you only display values at the latest batch). See https://keras.io/guides/customizing_what_happens_in_fit/#going-lowerlevel

Create them in `compile()`, add them as attributes to the model, and return them in a `metrics` property

File: examples/vision/nerf.py [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+
+
+test_imgs, test_rays = next(iter(train_ds))
+test_rays_flat, test_t_vals = test_rays
+
+loss_list = []
+
+
+class CustomCallback(keras.callbacks.Callback):
```
Comment (fchollet) [inline examples/vision/nerf.py:459]: Give it a descriptive name

File: examples/vision/nerf.py [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+nerf_model = get_nerf_model(num_layers=8, num_pos=num_pos)
+
+model = NeRF(nerf_model)
+model.compile(optimizer=keras.optimizers.Adam(), loss=keras.losses.MeanSquaredError())
+
+# Create a directory to save the images during training.
+if not os.path.exists("images"):
+    os.makedirs("images")
+
+model.fit(
```
Comment (fchollet) [inline examples/vision/nerf.py:497]: You should monitor generalization during `fit()` by using a separate `validation_data` dataset.

File: examples/vision/nerf.py [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+create_gif("images/*.png", "training.gif")
+
+"""
+## Visualize the training step
+
+Here we see the training step. With decreasing loss, the rendered
+image and the depth maps are getting better.
+"""
+
+Image(open("training.gif", "rb").read())
```
Comment (fchollet) [inline examples/vision/nerf.py:525]: This is jupyter-specific, pefer using matplotlib or similar to display the image so that the code works in a terminal

File: examples/vision/nerf.py [hunk 9]
```diff
@@ -0,0 +1,676 @@
+"""
+Title: 3-D Volumetric Rendering using NeRF
+Authors: [Aritra Roy Gosthipaty](https://twitter.com/arig23498), [Ritwik Raha](https://twitter.com/ritwik_raha)
+Date created: 2021/08/05
+Last modified: 2021/08/05
+Description: Minimal implementation of volumetric rendering as shown in NeRF.
+"""
+"""
+# Introduction
```
Comment (fchollet) [inline examples/vision/nerf.py:9]: Use `##` for all section titles, `#` is for the page title

File: examples/vision/nerf.py [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+    ax[1].imshow(keras.preprocessing.image.array_to_img(recons_img))
+    ax[1].set_title("Reconstructed")
+
+    ax[2].imshow(
+        keras.preprocessing.image.array_to_img(depth_map[..., None]), cmap="inferno"
+    )
+    ax[2].set_title("Depth Map")
+
+
+def get_trans_t(t):
```
Comment (fchollet) [inline examples/vision/nerf.py:568]: For the following functions, add one-line docstrings and use descriptive function names

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. Very cool example, very detailed and thorough. Great work!

**Comment by ariG23498** [comment]
@fchollet thanks for the thorough review. We have address all of the suggestions except one.
Please note the following:
- The `gif` and the `mp4` files follow different conventions for rendering in a notebook and in a terminal. Hence, we have shown the visulaization through a seperate uploaded file in markdown.
- We have kept the gif and mp4 generation code nevertheless for users to verify on their own.

Please let us know what you think about it!

File: examples/vision/nerf.py (hunk 5 above)
Comment (ariG23498) [reply examples/vision/nerf.py]: This was insightful. Thank you for your suggestion.

File: examples/vision/nerf.py [hunk 11]
```diff
@@ -0,0 +1,810 @@
...
+    .batch(BATCH_SIZE, drop_remainder=True, num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+"""
+## NeRF model
+
+The model is a multi-layer perceptron. It has relu as its non-linearity. 
+
+An exerpt from the paper:
```
Comment (fchollet) [inline examples/vision/nerf.py:393]: "excerpt". Tip: to catch typos, paste the example content into a Google doc and take a look at what gets highlighted.

File: examples/vision/nerf.py [hunk 12]
```diff
@@ -0,0 +1,810 @@
...
+
+"""
+## Training
+
+The training step is gathered in a `keras.Model` class so that we can
+make use of the `model.fit` call.
+"""
+
+loss_tracker = keras.metrics.Mean(name="loss")
+psnr_metric = keras.metrics.Mean(name="psnr")
```
Comment (fchollet) [inline examples/vision/nerf.py:490]: Prefer creating them in `__init__()` or `compile()` and attaching them to the model, to make the model self-contained

File: examples/vision/nerf.py [hunk 13]
```diff
@@ -0,0 +1,810 @@
...
+        # Get the images and the rays.
+        (images, rays) = inputs
+        (rays_flat, t_vals) = rays
+
+        with tf.GradientTape() as tape:
+            # Get the predictions from the model.
+            rgb, _ = render_rgb_depth(
+                model=self.nerf_model, rays_flat=rays_flat, t_vals=t_vals, rand=True
+            )
+            loss = self.loss(images, rgb)
```
Comment (fchollet) [inline examples/vision/nerf.py:513]: `loss = self.loss(...)` is confusing, prefer using `loss_fn` to make it clear that `self.loss_fn` is a function.

Alternatively you could use

`loss_value = self.loss(...)`

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Just minor comments.

File: examples/vision/nerf.py [hunk 14]
```diff
@@ -0,0 +1,810 @@
...
+
+            # apply positional encoding
+            positions.append(fn(2.0 ** i * x))
+
+    # return the positional encoding
+    return tf.concat(positions, axis=-1)
+
+
+def get_rays(height, width, focal, pose):
+    """
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/nerf.py]: Thank you for the detailed explanation.

File: examples/vision/nerf.py (hunk 11 above)
Comment (ariG23498) [reply examples/vision/nerf.py]: Neat trick. 😄

File: examples/vision/nerf.py (hunk 12 above)
Comment (ariG23498) [reply examples/vision/nerf.py]: Insightful. Sorry for not understanding this in the first pass.

File: examples/vision/nerf.py [hunk 15]
```diff
@@ -0,0 +1,778 @@
...
+
+In this example, we minimally implement the research paper
+[**NeRF: Representing Scenes as Neural Radiance Fields for View Synthesis**](https://arxiv.org/abs/2003.08934)
+by Ben Mildenhall et. al. The authors have proposed an ingenious way
+to **synthesise novel views** of a scene by modelling the *volumetric
+scene function* through a neural network.
+
+To construct an intuition, let me ask you a simple question. *What
+will happen if we input the coordinates of an image to a neural
+network and ask it to predict the color at that coordinate?*
```
Comment (fchollet) [inline examples/vision/nerf.py:19]: It isn't clear what the "coordinates of an image" means, please edit the text to make it clearer.

**Review by fchollet — COMMENTED** [review]
Looking very good, almost ready to merge. Please note that I've pushed copyedits, please pull them first.

File: examples/vision/nerf.py (hunk 15 above)
Comment (ariG23498) [reply examples/vision/nerf.py]: I have replaces the term `coordinates` with `pixel position`. Does that seem to make it clearer?

**Review by fchollet — APPROVED** [review]
Looking great! I've added a small copyedit touch.

Please add the generated files.

**Comment by ariG23498** [comment]
Hey @fchollet thanks for the copy edits. We have generated the files and added it here.

PS: The `autogen.py` script complained about the lines of the code. It expects less than 300 but our code was of 303 lines. We had to carefully shorten the code and apply black formatting to it (hence the number of commits).

**Comment by fchollet** [comment]
Thanks for the great contribution! I'll take it from here.
